addon/debounce-task.js

```diff
@@ -0,0 +1,148 @@
+import Ember from 'ember';
+import { run } from '@ember/runloop';
+import { assert } from '@ember/debug';
+import getOrAllocate from './utils/get-or-allocate';
+import { cancelTask } from './run-task';
+
+const { WeakMap } = Ember;
+
+/**
+ * A map of instances/debounce functions that allows us to
+ * store pending debounces per instance.
+ *
+ * @private
+ *
+ */
+const registeredDebounces = new WeakMap();
+
+/**
+   Runs the function with the provided name after the timeout has expired on the last
+   invocation. The timer is properly canceled if the object is destroyed before it is
+   invoked.
+
+   Example:
+
+   ```js
+   import Component from 'ember-component';
+   import { debounceTask, runDisposables } from 'ember-lifeline';
+
+   export default Component.extend({
+     logMe() {
+       console.log('This will only run once every 300ms.');
+     },
+
+     click() {
+       debounceTask(this, 'logMe', 300);
+     },
+
+     destroy() {
+       runDisposables(this);
+     }
+   });
+   ```
+
+   @method debounceTask
+   @param { Object } obj the instance to register the task for
+   @param { String } name the name of the task to debounce
+   @param { ...* } debounceArgs arguments to pass to the debounced method
+   @param { Number } wait the amount of time to wait before calling the method (in milliseconds)
+   @public
+   */
+export function debounceTask(obj, name, ...debounceArgs) {
+  assert(
+    `Called \`debounceTask\` without a string as the first argument on ${obj}.`,
+    typeof name === 'string'
+  );
+  assert(
+    `Called \`obj.debounceTask('${name}', ...)\` where 'obj.${name}' is not a function.`,
+    typeof obj[name] === 'function'
+  );
+  assert(
+    `Called \`debounceTask\` on destroyed object: ${obj}.`,
+    !obj.isDestroyed
+  );
+
+  let pendingDebounces = getOrAllocate(
```

Comment:
Does `getOrAllocate` actually provide value here? To my eyes it is more "magical" or "confusing" than the underlying implementation without it would be.

Consider:

```js
let pendingDebounces = getOrAllocate(
    registeredDebounces,
    obj,
    Object,
    getDebouncesDisposable
  );
```

versus:

```js
let pendingDebounces = registeredDebounces.get(obj);

if (!pendingDebounces) {
  pendingDebounces = {};
  registeredDebounces.set(obj, pendingDebounces);
  registerDisposable(obj, getDebouncesDisposable(pendingDebounces));
}
```

The second example is much clearer to me, and avoids passing the weakmap externally...

IMHO, we should remove the `getOrAllocate` utility...

Reply:
I could be persuaded, though it _is_ duplication of the same logic.

Reply:
I'm going to change it. I hate that signature.

Reply:
Inlined for `debounce-task` and `poll-task`. I created a module scoped function for `run-task`, as the same logic was shared 3 times in that module.